contracts/upgradeability/ClassicEternalStorageProxy.sol

```diff
@@ -0,0 +1,35 @@
+pragma solidity 0.4.19;
+
+import "./EternalStorage.sol";
+import "./OwnedUpgradeabilityProxy.sol";
+
+
+contract ClassicEternalStorageProxy is OwnedUpgradeabilityProxy, EternalStorage {
+
+    function () payable public {
+        address _impl = implementation();
+        require(_impl != address(0));
+        bytes4 sig;
+        assembly { sig := calldataload(0) }
+        uint256 len = getSize(sig);
+        assembly {
+            calldatacopy(0x0, 0x0, calldatasize)
+            let result := delegatecall(gas, _impl, 0x0, calldatasize, 0, len)
```

Comment:
I have noticed a mention of a bug related to `delegatecall` here: https://ropsten.etherscan.io/solcbuginfo (DelegateCallReturnValue). I also found the following post: https://github.com/paritytech/parity-ethereum/issues/6201#issuecomment-319294744.

So, since we are using zero pointer to get the return value I would make a test when we return some zero value from the method that is called by `delegatecall` to make sure that `result` will not be zero.
If it happens, probably we need to pass a free memory pointer instead of zero as the 5th parameter. So, I will leave the post https://ethereum.stackexchange.com/a/28457 here as reference that could help.

Reply:
I just tested adding two methods that returns `false` and `0`. Called them through `delegatecall`, on both cases the transaction call wasn't reverted and values were returned correctly